### PR TITLE
State transitions for a render pass

### PR DESCRIPTION
**Pull request by kvark**
This PR introduces a simple resource tracker, which is used to build the pipeline barriers gluing the main command buffer with the one composed for the render pass contents. Render passes are still not started properly, blocked on the W3C discussion about the parameters.

### CODE REVIEW

**Comment by kvark** [comment]
Note for the implementation side: we can't record the pass contents inline, because we need to do the transitions of all the resources before the pass starts, and we don't know those until we scan through the whole pass. That is why our approach is to record the render pass into a separate command buffer (thus the wgpu's command buffer ends up owning a list of native command buffers), and then the main command buffer gets appended with transition instructions.

Interestingly, we'd be able to record in-line if we had all the usage information upfront, and this can be organized in the remoting layer:
  - the client side could keep the trackers when going through the pass commands
  - client would send along the command list for a pass *together* with the tracker info
  - the server would then just record everything in-line, validating that the usage is correct

Naturally, this allows us to expose this to higher-level code, i.e. if WebGPU (native!) clients know the target usage, they can specify it up-front and get a slightly faster render pass recording.

**Comment by grovesNL** [comment]
Did you try updating the header with the bitflags change? IIRC it's still not generated correctly yet.

Could you explain how the native command buffers are synchronized?

> That is why our approach is to record the render pass into a separate command buffer (thus the wgpu's command buffer ends up owning a list of native command buffers), and then the main command buffer gets appended with transition instructions.

This seems significant, should we try to resolve this within WebGPU itself?

**Comment by kvark** [comment]
> Did you try updating the header with the bitflags change? IIRC it's still not generated correctly yet.

@grovesNL what's missing for us to get those bitflags going? I just saw the TODOs and assumed bitflags is the way forward. It would be sad if we can't use them, but not a showstopper.

I do confirm that the texture usage flags aren't getting generated by the cbindgen atm :/
Added a root Makefile to make testing those things easier.

> Could you explain how the native command buffers are synchronized?

Each WebGPU command buffer tracks the resource usage *locally* to it, knowing what state a resource needs to start with (this is still WIP, but the basic blocks are shaping up). At the same time, the device knows the state of all created resources in-between submissions. When a command buffer gets submitted, the device/queue introduces a "glue" command buffer to transit resources from the state they currently are into the state the next command buffer expects, before actually submitting that command buffer.

Now, a WebGPU command buffer consists of a sequence of native command buffers. Whenever there is a render/compute pass, the split happens, and it's recorded into a separate native command buffer. This is required, because we don't know the expected resource state at the start of a pass, and we can't transit resources while *in* the pass. So we record the pass contents into a different command buffer and record the transitions in the previous one. Those are then submitted as a sequence, simply (another option would be to submit the pass one as a secondary (SCB), but it doesn't seem necessary).

> This seems significant, should we try to resolve this within WebGPU itself?

I talked about this with @Kangz, and basically that's the only way to proceed for us if the usage is implicit. With explicit usage we could possibly do the transitions earlier and avoid multiple command buffers, but honestly I don't think it's a big deal in either the implementation or performance. We still submit all of them at once. Also, Google implementation doesn't care that much, since they aren't planning to record command buffers live (and defer it to submission time), and nobody else is working on an implementation, so we are kinda on our own here. If you feel strongly that we need explicit usage and 1:1 command buffer matching (between WebGPU and native), please speak up, the call is today.

**Comment by Kangz** [comment]
> If you feel strongly that we need explicit usage and 1:1 command buffer matching (between WebGPU and native)

As a data point, ANGLE's Vulkan backend didn't find performance issues when having one command buffer for the inside of each renderpass, and one command buffer threading them together. Both on Desktop and Android.

It would be possible to have an extension to do explicit barriers, but to be able to record a single command buffer live, you would need the barriers to contain both the "from" and the "to" state, which would be validated on submission.

**Comment by kvark** [comment]
@grovesNL please have a look at Makefile, the texture flags are blocked on https://github.com/eqrion/cbindgen/issues/232
I think we should proceed, and then when the bitflags are fully operational, we can polish the makefile more and enable it on CI.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	cargo build --manifest-path wgpu-native/Cargo.toml --features $(FEATURE_NATIVE)
+
+lib-rust: Cargo.lock wgpu-rs/Cargo.toml $(wildcard wgpu-rs/**/*.rs)
+	cargo build --manifest-path wgpu-rs/Cargo.toml --features $(FEATURE_RUST)
+
+wgpu-bindings/wgpu.h: Cargo.lock wgpu-bindings/src/*.rs lib-native
+	cargo +nightly run --manifest-path wgpu-bindings/Cargo.toml
+
+examples-native: lib-native wgpu-bindings/wgpu.h $(wildcard wgpu-native/**/*.c)
+	$(MAKE) -C examples
```
[the remark this answers is not in the document]
Comment (kvark) [reply Makefile]: I think we should, yeah

**Pull request merged**